File: backend/routers/gis.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from database import get_db
from models import Bus, Incident


router = APIRouter(
    prefix="/api/gis",
    tags=["GIS"]
)


# ============================================================
# GET INCIDENTS AS GEOJSON
# ============================================================

@router.get("/incidents")
def get_gis_incidents(
    db: Session = Depends(get_db)
):

    incidents = (
        db.query(Incident)
        .order_by(Incident.id.desc())
        .all()
    )

    features = []


    for incident in incidents:

        feature = {
            "type": "Feature",

            "geometry": {
                "type": "Point",

                # GeoJSON coordinates are always:
                # [longitude, latitude]
                "coordinates": [
                    incident.longitude,
                    incident.latitude
                ]
            },

            "properties": {
                "id": incident.id,

                "event_type": incident.event_type,

                "severity": incident.severity,

                "status": incident.status,

                "confirmation_count": (
                    incident.confirmation_count
                ),

                "unique_bus_count": (
                    incident.unique_bus_count
                ),

                "average_confidence": round(
                    incident.average_confidence,
                    4
                ),

                "verification_score": round(
                    incident.verification_score,
                    4
                ),

                "created_at": (
                    incident.created_at.isoformat()
                    if incident.created_at
                    else None
                ),

                "updated_at": (
                    incident.updated_at.isoformat()
                    if incident.updated_at
                    else None
                )
            }
        }

        features.append(feature)


    return {
        "type": "FeatureCollection",

        "features": features
    }


# ============================================================
# GET BUSES AS GEOJSON
# ============================================================

@router.get("/buses")
def get_gis_buses(
    db: Session = Depends(get_db)
):

    buses = (
        db.query(Bus)
        .order_by(Bus.bus_id)
        .all()
    )

    features = []


    for bus in buses:

        feature = {
            "type": "Feature",

            "geometry": {
                "type": "Point",

                # GeoJSON coordinates are always:
                # [longitude, latitude]
                "coordinates": [
                    bus.longitude,
                    bus.latitude
                ]
            },

            "properties": {
                "id": bus.id,

                "bus_id": bus.bus_id,

                "is_active": bus.is_active,

                "last_seen": (
                    bus.last_seen.isoformat()
                    if bus.last_seen
                    else None
                )
            }
        }

        features.append(feature)


    return {
        "type": "FeatureCollection",

        "features": features
    }

File: frontend/src/services/gis.ts
import { api } from './api';

import {
  buses,
  incidents,
} from '../mock/data';

import type {
  Bus,
  Incident,
} from '../types';


interface GeoJSONFeature {
  type: 'Feature';

  geometry: {
    type: 'Point';

    coordinates: [
      number,
      number
    ];
  };

  properties: Record<
    string,
    any
  >;
}


interface GeoJSONFeatureCollection {
  type: 'FeatureCollection';

  features: GeoJSONFeature[];
}


// ============================================================
// GIS INCIDENTS
// ============================================================

export async function getGISIncidents():
  Promise<Incident[]> {

  if (
    import.meta.env
      .VITE_USE_MOCKS !== 'false'
  ) {

    return Promise.resolve(
      incidents
    );

  }


  const { data } =
    await api.get<GeoJSONFeatureCollection>(
      '/api/gis/incidents'
    );


  return data.features.map(
    (feature) => {

      const properties =
        feature.properties;


      // GeoJSON format:
      // coordinates[0] = longitude
      // coordinates[1] = latitude

      const [
        longitude,
        latitude,
      ] =
        feature.geometry.coordinates;


      return {

        id:
          `INC-${String(
            properties.id
          ).padStart(
            3,
            '0'
          )}`,

        incident_id:
          properties.id,

        type:
          properties.event_type,

        severity:
          properties.severity,

        verification_score:
          (
            properties.verification_score ??
            properties.average_confidence ??
            0
          ) * 100,

        confirmed_buses:
          properties.unique_bus_count ??
          properties.confirmation_count ??
          0,

        status:
          properties.status,

        latitude,

        longitude,

        timestamp:
          properties.created_at,

        source_bus:
          properties.source_bus ??
          'UNKNOWN',

        frames_validated:
          properties.confirmation_count ??
          0,

        validation_total:
          4,

        description:
          `${String(
            properties.event_type ?? 'incident'
          )
            .replace(/_/g, ' ')
            .replace(
              /\b\w/g,
              (char: string) =>
                char.toUpperCase()
            )} detected by UrbanEye AI`,

      };

    }
  );

}


// ============================================================
// GIS BUSES
// ============================================================

export async function getBuses():
  Promise<Bus[]> {

  if (
    import.meta.env
      .VITE_USE_MOCKS !== 'false'
  ) {

    return Promise.resolve(
      buses
    );

  }


  const { data } =
    await api.get<GeoJSONFeatureCollection>(
      '/api/gis/buses'
    );


  return data.features.map(
    (feature) => {

      const properties =
        feature.properties;


      // GeoJSON format:
      // coordinates[0] = longitude
      // coordinates[1] = latitude

      const [
        longitude,
        latitude,
      ] =
        feature.geometry.coordinates;


      return {

        bus_id:
          properties.bus_id,

        latitude,

        longitude,

        status:
          properties.is_active
            ? 'ACTIVE'
            : 'OFFLINE',

        last_seen:
          properties.last_seen,

        recent_events:
          0,

        route:
          'Urban sensing route',

      };

    }
  );

}